### ISSUE TITLE
Not equal empty array reference with customParam

### ISSUE THREAD

**Issue by Kiiiwiii** [pbeshai/use-query-params#134]
### Reproduce:
1. have two query params, one is custom ArrayParam, another one is StringParam.
The custom param provider looks like this:
```
const commaArrayParams = {
  encode: (arr) => {
    if (arr.length === 0) {
      return undefined;
    }
    return arr.join(",");
  },
  decode: (str) => {
    if (!str || Array.isArray(str)) {
      return [];
    }

    return str.split(",");
  }
};
```

2. `useQueryParams` like this:
```
const config = {
  filters: withDefault(commaArrayParams, ["hi"]),
  sortBy: withDefault(StringParam, "a")
};

function App() {
  const [query, setQuery] = useQueryParams(config);
  ...
}
```

3. Toggle the sortBy query params by calling `setQuery({ sortBy: query.sortBy === "a" ? "b" : "a"})`.

### Expect
`query.filters` should not change.

### Actual
`query.filters` changed every time.

Sandbox to reproduce: https://codesandbox.io/s/admiring-dirac-fhv7o

## Some findings & Questions:

1. is it an expected behaviour?
2. when there is a non-value for `filters` in the url. `query.filters` not changes. (expected).
3. when change cached the returned empty array. so the return reference is the same. `query.filters` not changes.
```
const emptyArr = [];
const commaArrayParams = {
  encode: (arr) => {
    if (arr.length === 0) {
      return undefined;
    }
    return arr.join(",");
  },
  decode: (str) => {
    if (!str || Array.isArray(str)) {
      return emptyArr;
    }

    return str.split(",");
  }
};
```
Why this happened? 
When is the decoded function invoked?
Thanks for help.

**Comment by pbeshai** [comment]
This is due to your decode function I believe. `withDefault` only kicks in if a nully response is returned from decode. Does it work as you expect if you switch your commaArrayParams to be the following?

```
const commaArrayParams = {
  encode: (arr) => {
    if (arr.length === 0) {
      return undefined;
    }
    return arr.join(",");
  },
  decode: (str) => {
    if (!str || Array.isArray(str)) {
      return undefined; // <-- changed this
    }

    return str.split(",");
  }
};
```

In any case, I'd suggest using this approach for a comma delimited array encoding:

```js
import {
  encodeDelimitedArray,
  decodeDelimitedArray
} from "use-query-params";

/** Uses a comma to delimit entries. e.g. ['a', 'b'] => qp?=a,b */
const CommaArrayParam = {
  encode: (array) => encodeDelimitedArray(array, ','),
  decode: (arrayStr) => decodeDelimitedArray(arrayStr, ',')
};


const config = {
  filters: withDefault(CommaArrayParam, ["hello"]),
  sortBy: withDefault(StringParam, "asc")
};
```

**Comment by Kiiiwiii** [comment]
@pbeshai 
Thanks for your quick reply. It works with the codes you provided.
I'm still confused on one point. Could you please elaborate it a little bit?

I added some breakpoints in the example to see when the `decode function` is invoked. 
When in the url there is `?filters=hi`, and then I change another query param `sortBy`. The custom decode function for `query.filters` is not invoked. Thus `query.filters` keeps the same(expected). **So I guess you compare two strings from the url to decide whether to invoke the decode function?**

But then the confused part comes, when there is no `?filters=hi` in the url. The custom decode function for `query.filters` is invoked every time when `sortBy` changed. The decode function get `undefined` as input, and then returns `[]`. Every time `[]` is a new instance of array, that's why my useEffect console.log `query.filters` so often.
Change the return value to the same empty array instance or return `undefined`(which use the `withDefault` default value) will helps since the value is always the same. **But what I would expect in this case is, since the `filters value` from the url is not changing(`undefined` from both changes), it shouldn't fire the decode function for `query.filters`.** Isn't it??

Thank you!

**Comment by pbeshai** [comment]
I do think this may be a bug based on these lines: https://github.com/pbeshai/use-query-params/blob/master/src/useQueryParams.ts#L86-L88

```js
 // if we have a new encoded value, re-decode. otherwise reuse cache
    let encodedValue;
    let decodedValue;
    if (hasNewEncodedValue || encodedValuesCache[paramName] === undefined) {
      encodedValue = parsedQuery[paramName];
      decodedValue = paramConfig.decode(encodedValue);
    } else {
      encodedValue = encodedValuesCache[paramName];
      decodedValue = decodedValuesCache[paramName];
    }
```

Likely fixed if changed to:

```js
// if we have a new encoded value, re-decode. otherwise reuse cache
    let encodedValue;
    let decodedValue;
    if (
      hasNewEncodedValue ||
      (encodedValuesCache[paramName] === undefined &&
        decodedValuesCache[paramName] === undefined)
    ) {
      encodedValue = parsedQuery[paramName];
      decodedValue = paramConfig.decode(encodedValue);
    } else {
      encodedValue = encodedValuesCache[paramName];
      decodedValue = decodedValuesCache[paramName];
    }
```

The check of `encodedValuesCache[paramName] === undefined` is done to see if we have ever encoded before, but it's possible encoded = undefined and decoded is a different value as you've shown in your example. If they both returned undefined, it still would call the decode function each time, but presumably this is low cost and wouldn't break any equality down the line.

If you want to add a test for your use case (probably easiest via components-test.tsx) and update the clause above and verify it fixes it, we can patch the lib.

**Comment by Kiiiwiii** [comment]
@pbeshai thanks for the explanation. I will try to do it on the weekend

**Comment by pbeshai** [comment]
Published in v1.1.9
